### PR TITLE
Deep Lake upgrade to include attribute search, distance metrics, returning scores and MMR

### PR DESCRIPTION
**Pull request by davidbuniat**
### Features include

- Metadata based embedding search
- Choice of distance metric function (`L2` for Euclidean, `L1` for Nuclear, `max` L-infinity distance, `cos` for cosine similarity, 'dot' for dot product. Defaults to `L2`
- Returning scores
- Max Marginal Relevance Search
- Deleting samples from the dataset

### Notes
- Added numerous tests, let me know if you would like to shorten them or make smarter

### CODE REVIEW

**Comment by davidbuniat** [comment]
kk I see mypy issues now, will fix them in a bit.
```
poetry run mypy .
langchain/vectorstores/deeplake.py:56: error: Item "None" of "Optional[Dict[str, str]]" has no attribute "items"  [union-attr]
langchain/vectorstores/deeplake.py:155: error: Name "metadatas" already defined on line 126  [no-redef]
langchain/vectorstores/deeplake.py:157: error: Argument 3 to "zip" has incompatible type "Optional[List[Dict[Any, Any]]]"; expected "Iterable[Dict[Any, Any]]"  [arg-type]
langchain/vectorstores/deeplake.py:174: error: Signature of "similarity_search" incompatible with supertype "VectorStore"  [override]
langchain/vectorstores/deeplake.py:174: note:      Superclass:
langchain/vectorstores/deeplake.py:174: note:          def similarity_search(self, query: str, k: int = ..., **kwargs: Any) -> List[Document]
langchain/vectorstores/deeplake.py:174: note:      Subclass:
langchain/vectorstores/deeplake.py:174: note:          def similarity_search(self, query: Optional[str] = ..., embedding: Optional[Embeddings] = ..., k: int = ..., distance_metric: Optional[str] = ..., use_maximal_marginal_relevance: Optional[bool] = ..., fetch_k: Optional[int] = ..., filter: Optional[Dict[str, str]] = ..., return_score: Optional[bool] = ..., **kwargs: Any) -> Any
langchain/vectorstores/deeplake.py:206: error: Incompatible types in assignment (expression has type "partial[bool]", variable has type "Optional[Dict[str, str]]")  [assignment]
langchain/vectorstores/deeplake.py:222: error: Incompatible types in assignment (expression has type "Union[Embeddings, List[float]]", variable has type "Optional[Embeddings]")  [assignment]
langchain/vectorstores/deeplake.py:222: error: Argument 1 to "embed_query" of "Embeddings" has incompatible type "Optional[str]"; expected "str"  [arg-type]
langchain/vectorstores/deeplake.py:226: error: Argument "k" to "search" has incompatible type "Optional[int]"; expected "int"  [arg-type]
langchain/vectorstores/deeplake.py:227: error: Item "None" of "Optional[str]" has no attribute "lower"  [union-attr]
langchain/vectorstores/deeplake.py:245: error: List comprehension has incompatible type List[Tuple[Document, float]]; expected List[Document]  [misc]
langchain/vectorstores/deeplake.py:249: error: Signature of "similarity_search_by_vector" incompatible with supertype "VectorStore"  [override]
langchain/vectorstores/deeplake.py:249: note:      Superclass:
langchain/vectorstores/deeplake.py:249: note:          def similarity_search_by_vector(self, embedding: List[float], k: int = ..., **kwargs: Any) -> List[Document]
langchain/vectorstores/deeplake.py:249: note:      Subclass:
langchain/vectorstores/deeplake.py:249: note:          def similarity_search_by_vector(self, embedding: List[float], distance_metric: Optional[str] = ..., k: int = ..., filter: Optional[Dict[str, str]] = ...) -> List[Document]
langchain/vectorstores/deeplake.py:266: error: Argument "embedding" to "similarity_search" of "DeepLake" has incompatible type "List[float]"; expected "Optional[Embeddings]"  [arg-type]
langchain/vectorstores/deeplake.py:289: error: Signature of "max_marginal_relevance_search_by_vector" incompatible with supertype "VectorStore"  [override]
langchain/vectorstores/deeplake.py:289: note:      Superclass:
langchain/vectorstores/deeplake.py:289: note:          def max_marginal_relevance_search_by_vector(self, embedding: List[float], k: int = ..., fetch_k: int = ...) -> List[Document]
langchain/vectorstores/deeplake.py:289: note:      Subclass:
langchain/vectorstores/deeplake.py:289: note:          def max_marginal_relevance_search_by_vector(self, embedding: List[float], distance_metric: Optional[str] = ..., k: int = ..., fetch_k: int = ..., filter: Optional[Dict[str, str]] = ...) -> List[Document]
langchain/vectorstores/deeplake.py:309: error: Argument "embedding" to "similarity_search" of "DeepLake" has incompatible type "List[float]"; expected "Optional[Embeddings]"  [arg-type]
langchain/vectorstores/deeplake.py:311: error: Signature of "max_marginal_relevance_search" incompatible with supertype "VectorStore"  [override]
langchain/vectorstores/deeplake.py:311: note:      Superclass:
langchain/vectorstores/deeplake.py:311: note:          def max_marginal_relevance_search(self, query: str, k: int = ..., fetch_k: int = ...) -> List[Document]
langchain/vectorstores/deeplake.py:311: note:      Subclass:
langchain/vectorstores/deeplake.py:311: note:          def max_marginal_relevance_search(self, query: str, distance_metric: Optional[str] = ..., k: int = ..., fetch_k: int = ..., filter: Optional[Dict[str, str]] = ...) -> List[Document]
langchain/vectorstores/deeplake.py:406: error: Incompatible types in assignment (expression has type "partial[bool]", variable has type "Optional[Dict[str, str]]")  [assignment]
langchain/vectorstores/deeplake.py:411: error: Argument 1 to "sorted" has incompatible type "Optional[List[str]]"; expected "Iterable[str]"  [arg-type]
langchain/vectorstores/deeplake.py:4[13](https://github.com/hwchase17/langchain/actions/runs/4622552973/jobs/8175375996?pr=2455#step:6:14): error: Argument 1 to "len" has incompatible type "Optional[List[str]]"; expected "Sized"  [arg-type]
tests/integration_tests/vectorstores/test_deeplake.py:11: error: Function is missing a return type annotation  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:21: error: Function is missing a type annotation  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:82: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:90: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:99: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:111: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:120: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:138: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
tests/integration_tests/vectorstores/test_deeplake.py:[14](https://github.com/hwchase17/langchain/actions/runs/4622552973/jobs/8175375996?pr=2455#step:6:15)8: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
Found [27](https://github.com/hwchase17/langchain/actions/runs/4622552973/jobs/8175375996?pr=2455#step:6:28) errors in 2 files (checked 555 source files)
make: *** [Makefile:[31](https://github.com/hwchase17/langchain/actions/runs/4622552973/jobs/8175375996?pr=2455#step:6:32): lint] Error 1
```

File: langchain/vectorstores/deeplake.py [hunk 1]
```diff
@@ -3,40 +3,72 @@
...
 
-    # Sort the distances and return the indices of the k nearest vectors
-    nearest_indices = np.argsort(distances)[:k]
-    return nearest_indices.tolist()
+def dp_filter(x: dict, filter: Optional[Dict[str, str]] = None) -> bool:
+    """Filter helper function for Deep Lake"""
+    metadata = x['metadata'].data()["value"]
+    if any([str(v) != str(metadata[k]) if k in metadata else True for k, v in filter.items()]):
+        return False
+    return True
```
Comment (farizrahman4u) [inline langchain/vectorstores/deeplake.py:58]: Refac:

(Also, why str conversion?)
```suggestion
def dp_filter(x: Dict, filter: Optional[Dict[str, str]] = None) -> bool:
    """Filter helper function for Deep Lake"""
    metadata = x['metadata'].data()["value"]
    return all(k in metadata and str(v) == str(metadata[k]) for k, v in filter.items())
```

**Review by farizrahman4u — APPROVED** [review]

File: langchain/vectorstores/deeplake.py (hunk 1 above)
Comment (davidbuniat) [reply langchain/vectorstores/deeplake.py]: Good point!

**Comment by davidbuniat** [comment]
@hwchase17 should be good to go from our side! Let us know if there is anything we need to improve.

**Review by hwchase17 — APPROVED** [review]
this is great! will merge this in now

i do think one good improvement would be to show how to construct a retriever object that utilizes these methods

**Comment by davidbuniat** [comment]
Next on our todo!

**Comment by happy15** [comment]
Thanks for the work. Just wonder when will this be released out for usage via pip upgrade?

**Comment by davidbuniat** [comment]
@happy15 I believe it just got released with `0.0.134`, let me know if you have any feedback or questions. Would love to see if any specific feature you would need for Deep Lake integration.

**Pull request merged**